### PR TITLE
Support more TE configurations

### PR DESCRIPTION
**Pull request by zlsh80826**
Added supported features
- RNSNorm
- Activations (ReLU ...), gated activations
- GQA
- RoPE
- Sequence Parallel
- non-causal masking
- bias

### CODE REVIEW

**Comment by nouiz** [comment]
This depend on https://github.com/NVIDIA/TransformerEngine/pull/602

**Review by mingxu1067 — APPROVED** [review]

File: praxis/contrib/gpu/scripts_gpu/te_helper.py [hunk 1]
```diff
@@ -112,19 +115,85 @@ def _fprop(
...
+            # TODO(rewang): rope
+            if issubclass(attn_tpl.cls, attentions.DotProductAttention):
+                # Check the DotProductAttention parameters are aligned to TE's attention
+                assert attn_tpl.internal_enable_query_scale or attn_tpl.scale_logits_by_head_dims
+                assert not (attn_tpl.internal_enable_query_scale and attn_tpl.scale_logits_by_head_dims)
+                assert not attn_tpl.internal_enable_per_dim_scale
+                assert not attn_tpl.scale_query_by_dim_per_head
+                assert not attn_tpl.dconv_qkv
+                assert not attn_tpl.internal_gshard_gaussian_init
+                assert not attn_tpl.use_rotary_position_emb
```
Comment (hx89) [inline praxis/contrib/gpu/scripts_gpu/te_helper.py]: Should we remove this assert? Since DotProductAttention can also use RoPE: https://github.com/google/saxml/blob/main/saxml/server/pax/lm/params/lm_cloud.py#L145

Comment (zlsh80826) [reply praxis/contrib/gpu/scripts_gpu/te_helper.py]: Hi @hx89, do you have a usecase that using RoPE with `DotProductAttention` module? The [original request of RoPE](https://github.com/google/paxml/blob/main/paxml/tasks/lm/params/nvidia.py#L509-L517) is paired with `GroupedQueryAttention` module, this is the reason I didn't enable it. I can enable it if we have a real case for it.

Comment (zlsh80826) [reply praxis/contrib/gpu/scripts_gpu/te_helper.py]: Offline discussed with Anna and remove the assertion.

File: praxis/contrib/gpu/scripts_gpu/te_helper.py [hunk 2]
```diff
@@ -112,19 +116,90 @@ def _fprop(
...
+                assert not (attn_tpl.internal_enable_query_scale and attn_tpl.scale_logits_by_head_dims)
+                assert not attn_tpl.internal_enable_per_dim_scale
+                assert not attn_tpl.scale_query_by_dim_per_head
+                assert not attn_tpl.dconv_qkv
+                assert not attn_tpl.internal_gshard_gaussian_init
+                assert attn_tpl.relative_bias_tpl is None
+                assert attn_tpl.attention_extra_logit is None
+                assert attn_tpl.ngrammer_tpl is None
+                te_tpl.enable_rotary_pos_emb = attn_tpl.use_rotary_position_emb
+                if issubclass(attn_tpl.rotary_position_emb_tpl, embedding_softmax.RotaryPositionalEmbedding):
```
Comment (ashors1) [inline praxis/contrib/gpu/scripts_gpu/te_helper.py]: I think this should be 
```
if issubclass(attn_tpl.rotary_position_emb_tpl, embedding_softmax.RotaryPositionalEmbedding.cls):
```

File: praxis/contrib/gpu/scripts_gpu/te_helper.py [hunk 3]
```diff
@@ -112,19 +116,90 @@ def _fprop(
...
+            elif issubclass(attn_tpl.cls, grouped_query_attention.GroupedQueryAttention):
+                te_tpl.num_gqa_groups = attn_tpl.num_kv_heads
+                if attn_tpl.rope_min_max_timescales is not None:
+                    te_tpl.enable_rotary_pos_emb = True
+                    te_tpl.rotary_pos_emb_windows = attn_tpl.rope_min_max_timescales
+                assert attn_tpl.atten_temp == 1.
+            elif issubclass(attn_tpl.cls, multi_query_attention.MultiQueryDotProductAttention):
+                te_tpl.num_gqa_groups = attn_tpl.num_kv_heads
+                te_tpl.enable_rotary_pos_emb = attn_tpl.use_rotary_position_emb
+                if issubclass(attn_tpl.rotary_position_emb_tpl, embedding_softmax.RotaryPositionalEmbedding):
```
Comment (ashors1) [inline praxis/contrib/gpu/scripts_gpu/te_helper.py]: Same here

File: praxis/contrib/gpu/scripts_gpu/te_helper.py (hunk 2 above)
Comment (zlsh80826) [reply praxis/contrib/gpu/scripts_gpu/te_helper.py]: Fixed, thanks for catching that

**Review by ashors1 — APPROVED** [review]
LGTM! I will manually pull these changes into the `patch/te_support` branch

**Comment by zlsh80826** [comment]
Closed as it is cherry-picked to https://github.com/nvjax-svc-0/praxis/commits/patch/te_support.
